sled-storage/src/manager.rs

```diff
@@ -0,0 +1,1034 @@
+        let resources = StorageResources::default();
+        let (update_tx, update_rx) = watch::channel(resources.clone());
+        (
+            StorageManager {
+                log: log.new(o!("component" => "StorageManager")),
+                state: StorageManagerState::WaitingForKeyManager,
+                tx: tx.clone(),
+                rx,
+                resources,
+                queued_u2_drives: HashSet::new(),
+                key_requester,
+                resource_updates: update_tx,
+                last_logged_capacity: QUEUE_SIZE,
+            },
+            StorageHandle { tx, resource_updates: update_rx },
+        )
+    }
+
+    /// Run the main receive loop of the `StorageManager`
+    ///
+    /// This should be spawned into a tokio task
+    pub async fn run(mut self) {
+        loop {
+            const QUEUED_DISK_RETRY_TIMEOUT: Duration = Duration::from_secs(10);
+            let mut interval = interval(QUEUED_DISK_RETRY_TIMEOUT);
+            interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+            tokio::select! {
+                res = self.step() => {
+                    if let Err(e) = res {
+                        warn!(self.log, "{e}");
+                    }
+                }
+                _ = interval.tick(),
+                    if self.state == StorageManagerState::QueueingDisks =>
+                {
+                    if self.add_queued_disks().await {
+                        let _ = self.resource_updates.send_replace(self.resources.clone());
+                    }
+                }
+            }
+        }
+    }
+
+    /// Process the next event
+    ///
+    /// This is useful for testing/debugging
+    pub async fn step(&mut self) -> Result<(), Error> {
+        const CAPACITY_LOG_THRESHOLD: usize = 10;
+        // We check the capacity and log it every time it changes by at least 10
+        // entries in either direction.
+        let current = self.tx.capacity();
+        if self.last_logged_capacity.saturating_sub(current)
+            >= CAPACITY_LOG_THRESHOLD
+        {
+            info!(
+                self.log,
+                "Channel capacity decreased";
+                "previous" => ?self.last_logged_capacity,
+                "current" => ?current
+            );
+            self.last_logged_capacity = current;
+        } else if current.saturating_sub(self.last_logged_capacity)
+            >= CAPACITY_LOG_THRESHOLD
+        {
+            info!(
+                self.log,
+                "Channel capacity increased";
+                "previous" => ?self.last_logged_capacity,
+                "current" => ?current
+            );
+            self.last_logged_capacity = current;
+        }
+        // The sending side never disappears because we hold a copy
+        let req = self.rx.recv().await.unwrap();
+        info!(self.log, "Received {:?}", req);
+        let should_send_updates = match req {
```

Comment:
This update organization here makes a lot of sense to me - it's clear, and they propagate in one spot.